libraries/ESP8266WebServer/src/ESP8266WebServerSecure.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+  ESP8266WebServerSecure.cpp - Dead simple HTTPS web-server.
+  Supports only one simultaneous client, knows how to handle GET and POST.
+
+  Copyright (c) 2017 Earle F. Philhower, III. All rights reserved.
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+  Modified 8 May 2015 by Hristo Gochkov (proper post and file upload handling)
+*/
+
+
+#include <Arduino.h>
+#include <libb64/cencode.h>
+#include "WiFiServer.h"
+#include "WiFiClient.h"
+#include "ESP8266WebServerSecure.h"
+
+
+ESP8266WebServerSecure::ESP8266WebServerSecure(IPAddress addr, int port) 
+: _serverSecure(addr, port)
+{
+}
+
+ESP8266WebServerSecure::ESP8266WebServerSecure(int port)
+: _serverSecure(port)
+{
+}
+
+void ESP8266WebServerSecure::setServerKeyAndCert_P(const uint8_t *key, int keyLen, const uint8_t *cert, int certLen)
+{
+    _serverSecure.setServerKeyAndCert_P(key, keyLen, cert, certLen);
+}
+
+void ESP8266WebServerSecure::setServerKeyAndCert(const uint8_t *key, int keyLen, const uint8_t *cert, int certLen)
+{
+    _serverSecure.setServerKeyAndCert(key, keyLen, cert, certLen);
+}
+
+ESP8266WebServerSecure::~ESP8266WebServerSecure() {
+  // Nothing to do here.
+  // Base class's destructor will be called to clean up itself
+}
+
+// We need to basically cut-n-paste these from WebServer because of the problem
```

Comment:
This makes no sense to me. Where is there an assignment such as ESP8266Webserver = ESP8266WebServerSecure? If there is anywhere where an assignment is done, then there is likely a bug there.

Reply:
The problem is not the webserver, but the WiFiServer.available() model...
````
...
    WiFiClient(Secure) client = _server(Secure).available();
...
````
Because WiFiServer and it's children return a copy of a object, not a pointer to a new()'d one, the caller needs to know how big it will be/etc.  So the WebServerSecure class can't fall back on the WebServer::handle() method and let the vtable work its magic.  Object slicing, if I remember the term properly.

If there was an method like "virtual WiFiClient *WiFiServer.available();" it would be easier to reuse the WebServer code.  The Secure HTTPS Updater example shows this use.

Reply:
I know about object slicing. When you encounter it, it's usually a sign of a bug or of bad design (not always).
I should investigate this. Sigh, so much to do!
In any case, not a fault in your implementation, which is why I only commented and approved.

Reply:
This can also be solved with the template-based implementation and passing by reference, to some extent. Plus a few helper methods to clear the current client and get a new one from the WebServer. Will handle in another PR.